metatron/src/main/java/uk/gov/ida/eidas/metatron/domain/MetadataResolverService.java

```diff
@@ -1,39 +1,74 @@
+import org.opensaml.saml.saml2.metadata.AssertionConsumerService;
+import org.opensaml.saml.saml2.metadata.EntityDescriptor;
+import org.opensaml.saml.saml2.metadata.SPSSODescriptor;
+import org.opensaml.saml.security.impl.MetadataCredentialResolver;
+import org.opensaml.security.credential.Credential;
+import org.opensaml.security.credential.UsageType;
+import org.opensaml.security.criteria.UsageCriterion;
+import org.opensaml.security.x509.X509Credential;
+import uk.gov.ida.eidas.metatron.exceptions.MetatronClientException;
+import uk.gov.ida.eidas.metatron.exceptions.MetatronServerException;
+import uk.gov.ida.notification.contracts.CountryMetadataResponse;
 import uk.gov.ida.saml.metadata.ExpiredCertificateMetadataFilter;
+import uk.gov.ida.saml.metadata.JerseyClientMetadataResolver;
 import uk.gov.ida.saml.metadata.PKIXSignatureValidationFilterProvider;
+import uk.gov.ida.saml.metadata.factories.CredentialResolverFactory;
 import uk.gov.ida.saml.metadata.factories.MetadataResolverFactory;
 
 import javax.ws.rs.client.Client;
+import javax.xml.namespace.QName;
+import java.net.URI;
+import java.security.cert.CertificateEncodingException;
 import java.util.ArrayList;
-import java.util.HashMap;
+import java.util.Base64;
 import java.util.List;
 import java.util.Map;
+import java.util.function.Function;
 import java.util.stream.Collectors;
 
 public class MetadataResolverService {
 
     protected static long MIN_REFRESH_DELAY_MS = 60_000;
     protected static long MAX_REFRESH_DELAY_MS = 600_000;
 
-    private Map<String, MetadataResolver> metadataResolverMap;
-    private ExpiredCertificateMetadataFilter expiredCertificateMetadataFilter;
-    private MetadataResolverFactory metadataResolverFactory;
+    private final Map<EidasCountryConfig, MetadataResolver> metadataResolverMap;
+    private final ExpiredCertificateMetadataFilter expiredCertificateMetadataFilter;
+    private final MetadataResolverFactory metadataResolverFactory;
+    private final CredentialResolverFactory credentialResolverFactory;
 
-    public MetadataResolverService(EidasConfig countriesConfig, MetadataResolverFactory metadataResolverFactory) {
+    public MetadataResolverService(
+            EidasConfig countriesConfig,
+            MetadataResolverFactory metadataResolverFactory,
+            CredentialResolverFactory credentialResolverFactory) {
         this.metadataResolverFactory = metadataResolverFactory;
-        this.metadataResolverMap = new HashMap<>();
+        this.credentialResolverFactory = credentialResolverFactory;
         this.expiredCertificateMetadataFilter = new ExpiredCertificateMetadataFilter();
-
-        metadataResolverMap = countriesConfig.getCountries().stream().collect(
-                Collectors.toMap(EidasCountryConfig::getEntityId, this::createMetadataResolver));
+        this.metadataResolverMap = countriesConfig.getCountries().stream()
+                .collect(Collectors.toMap(Function.identity(), this::createMetadataResolver));
```

Comment:
What's the benefit of having the key be the actual config, rather than the entity id? It seems it means implementing the logic to initially get the config out of the map by iterating through all the keys and checking against a given entity id. And once we've done that once we use that config twice more to access the map again to get the metadata resolvers.
It seems like it might be easier to just initially access it via the entity id, and then just call `getEntityId()` on that config when we need to fetch the metadata from the map in future.

Reply:
I'm with Chris on this one - it seems like over-egging the pudding a bit.  If I've understood this correctly, this is a way of getting the country code and the entityId to put into the CountryMetadataResponse but I really feel that the keys should have a single responsibility of being a key; it might be a clearer statement of intent to have an object composed of the properties we want and the metadata resolver as value; we'd retain the advantages of a map that way too.  I have crazy ideas about CountryMetadataResponses that contain Functions and Suppliers that get called by the getters but I've just said we should probably keep it simple so I won't suggest them!

Reply:
Addressed.